### DISCUSSION TITLE
Is there no .isConencted() when using JedisCluster?

### DISCUSSION THREAD

**Discussion by thxrben** [xetorthio/jedis#3298] [General]
I am getting no connection currently in one of my projects, therefore I wanted to see if the connection to the cluster is actuall "up".

I installed redis-cli on the clients where my project runs on and made sure that my cluster works aswell as the connection to it. And well, it does.
However, seemingly, any call to modify sets (in my case) does result in an error (No more cluster attempts left -> Failed to connect to any host for DNS name -> ConnectionTimeout).

First of all, I am not using any DNS names/hostnames, but only the raw IPs and Ports.

I have seen #2225, but I dont get any exception.

Also, I just noticed the getConnectionFromSlot() function. Do i have to iterate over all 16k Slots to check if all connections are up / if I can make requests?
Or should I use the getClusterNodes() function to retrieve a Map? Also, what is the String Key in the map?

**Comment by thxrben** [comment]
Using this code, I am trying to connect, but the first exception thrown is "Failed to connect to any host resolved for DNS name" at my System.out.println where i print the key and then pool.isClosed() etc.

```
        Set<HostAndPort> redisClusterNodes = new HashSet<>();

        try {
            redisClusterNodes.add(new HostAndPort("192.168.100.110", 7000));
            redisClusterNodes.add(new HostAndPort("192.168.100.110", 7001));
            redisClusterNodes.add(new HostAndPort("192.168.100.110", 7002));
            redisClusterNodes.add(new HostAndPort("192.168.100.110", 7003));

            redisClusterNodes.add(new HostAndPort("192.168.100.111", 7000));
            redisClusterNodes.add(new HostAndPort("192.168.100.111", 7001));
            redisClusterNodes.add(new HostAndPort("192.168.100.111", 7002));
            redisClusterNodes.add(new HostAndPort("192.168.100.111", 7003));
            //redisClusterNodes.add("192.168.1.110", 7000);
            cluster = new JedisCluster(redisClusterNodes);
        } catch (Exception e) {
            e.printStackTrace();
        }

        Map<String, ConnectionPool> nodes = cluster.getClusterNodes();
        Iterator<Map.Entry<String, ConnectionPool>> it = nodes.entrySet().stream().iterator();
        System.out.println("### REDIS CLUSTER CONNECTION POOL ###\n");
        while(it.hasNext()) {
            Map.Entry<String, ConnectionPool> entry = it.next();
            ConnectionPool pool = entry.getValue();
            if(pool != null) {
                System.out.println("Key: "+entry.getKey()+" | Closed? "+pool.isClosed()+" - Resource.isConnected? "+pool.getResource().isConnected()+" - Resource.isBroken? "+pool.getResource().isBroken());
            }
        }
```

This cant be right?

**Comment by sazzad16** [comment]
Your Java application can't connect to at least one of the cluster nodes, probably due to firewall or something else.

Split your `System.out.println` so that you can find which node it is. E.g.

```
System.out.println("Key: "+entry.getKey());
System.out.println("Closed? "+pool.isClosed());
System.out.println("Resource.isConnected? "+pool.getResource().isConnected()+" - Resource.isBroken? "+pool.getResource().isBroken());
```

---

Lastly,

> what is the String Key in the map?

`<HOST>:<PORT>`, the address of the node.

**Comment by thxrben** [comment]
Oh. I think I can see the issue. However it does not make any sense to me:

I've split the lines according to you.
It now prints this as the first item of the array:

Key: 192.168.1.111:7002
Closed? false
redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.
...



I am using a IPSec Tunnel between the Client and the Redis-Cluster. 
Since both Client & Server are on the same subnet (192.168.1.0/24), I have to perform BiNat, so that the server is rechable on 192.168.100.xxx, and the clients are reachable by the servers at 192.168.200.xxx.

Now, since this is a Redis-Cluster, I can imagine that if it tries to e.g. read, the Cluster says "nope, I dont have that key/set available, please go to instance Y".
However, even then it doesnt make sense to me, since I provided all hosts.

Does REDIS upon connection use any sort of hostnames? The "DNS" part is still unclear to me.

**Reply by thxrben** [reply]
Addition: The Error is thrown in the line with the pool.getResource()

**Comment by thxrben** [comment]
Bump

**Comment by sazzad16** [comment]
> The "DNS" part is still unclear to me.

The host you provide goes though JVM's DNS. You don't have to be concerned when it's an IP because the [JavaDoc](https://docs.oracle.com/javase/7/docs/api/java/net/InetAddress.html#getAllByName(java.lang.String)) says:

_"If a literal IP address is supplied, only the validity of the address format is checked."_

> Does REDIS upon connection use any sort of hostnames?

No.

> ... since I provided all hosts.

It doesn't matter whether you all provide all hosts or not. Jedis uses the hosts reported by Redis server, for internal cluster management, not provided by you. The hosts provided are used as the _first contact point_.

---

When asking for help, it's better to provide a longer stacktrace instead of only the message. I believe your exception contains another suppressed exception as well. Whether the exception contains another suppressed one or not makes a huge difference while going deep into the issue. Because when not given, our first assumption would be that there's no suppressed one.

Moreover, your actual cluster node addresses are not the same as the ones you're providing is another huge information that was missing.

Please try to provide as much information as you can as early as possible, of course in a sanely concise way.

Please don't take this comment in a negative way. I appreciate you created a discussion instead of an issue which most users don't follow even after requests. And of course thank you for using Jedis.

---

Now, coming to the solution, as your actual cluster node addresses different from the addresses your java application (Jedis) can communicate with, provide an implementation of `HostAndPortMapper` in the cluster constructor. I think in your case, upon receiving a `192.168.1.*` IP, it would return a `192.168.100.*` IP.

**Comment by thxrben** [comment]
Sorry for not being more clear in my question. I wasn't aware that Redis responds with IP-Addresses to tell where the requested data can be fetched.

For me it seems like its just an Interface with "no" implementations? In the means of its nowhere used inside the library?
I was only able to find constructors in "JedisClusterInfoCache" and not in the JedisCluster class?

I'd really like to use it. Am I missing documentation?
If there's none, I'd be happy to help documenting the code, but for now I cannot seem to find the correct Constructor (if there is any).

**Comment by sazzad16** [comment]
There are no implementations of `HostAndPortMapper` in main source code. There are a few implementations in test source; for example in [SSLJedisClusterTest](https://github.com/redis/jedis/blob/master/src/test/java/redis/clients/jedis/SSLJedisClusterTest.java) and [SSLJedisSentinelPoolTest](https://github.com/redis/jedis/blob/master/src/test/java/redis/clients/jedis/SSLJedisSentinelPoolTest.java).

Any form of contribution is always welcome, even documentation!

I just realized there are no JedisCluster constructor taking a HostAndPortMapper directly. You may choose any constructor taking `JedisClientConfig` where you'd be able to pass your own HostAndPortMapper.

**Comment by thxrben** [comment]
Jep, this worked out.
I will see if I can go and document some stuff.

Just for anyone else who is seeing this:

Incase of my 2 Redis-Servers with multiple Instances, this was enough:

```
    private static final HostAndPortMapper hostAndPortMap = (HostAndPort hostAndPort) -> {
        String host = hostAndPort.getHost();
        int port = hostAndPort.getPort();
        if (host.equals("192.168.1.110")) {
            host = "192.168.100.110";
        }
        if (host.equals("192.168.1.111")) {
            host = "192.168.100.111";
        }
        return new HostAndPort(host, port);
    };
```

You can certainly create a Hashmap that maps the IPs 192.168.1.110 to 192.168.100.110 and just look it up inside that function, but since i only have 2 of them, I just used an If-Statement.

I supplied it to the JedisCluster Constructor like this:

```
new JedisCluster(redisClusterNodes, DefaultJedisClientConfig.builder()
                    .hostAndPortMapper(hostAndPortMap).build());
```

**Comment by ggivo** [comment]
Hi, @thxrben 
Thanks for bringing this to the team's attention.

From the latest comments, my understanding is that the initially reported issue is resolved. 
I have opened a dedicated issue to track [documentation improvements of `HostAndPortMapper`](https://github.com/redis/jedis/issues/4112). As always, any contribution is highly appreciated.

**Reply by thxrben** [reply]
Yes, the issue is resolved. I should've marked it as such. Thanks for your work!
